### PR TITLE
refactor: rebuild the compiler

### PR DESCRIPTION
**Pull request by KnisterPeter**
The first compiler design has had a lot of flaws from architectural
perspective. This was rebuild with this commit.
It is working now without changes for the 0.x version of the runtime so
a migration would be possible.

### CODE REVIEW

File: tests/compiler/cli/test.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  );
+
+  expect(result.stdout).toBe(`// tslint:disable
+import { TSDI } from 'tsdi';
+import {
+  Container as C1_Container,
+  Dependency as C3_Dependency,
+  Test as C2_Test
+} from '../container';
+export class ContainerImpl extends C1_Container {
```
Comment (Prior99) [inline tests/compiler/cli/test.ts]: I understand what this is needed for, but my guess would be that one will afterwards at project's toplevel layer import the `ContainerImpl`s from the generated code and instanciate them.

If you'd have some kind of static registry in tsdi (`ContainerRegistry.implFor(Container)`) we could avoid to force the user into a renaming.

The container impl's would self-register themselfs at the registry in the generated code below (`ContainerRegistry.register(ContainerImpl)`).

Just a rough idea.

File: lib/compiler/decorators.ts [hunk 2]
```diff
@@ -1,26 +1,24 @@
+import { addTsdiMarker } from '../marker';
```
Comment (Prior99) [inline lib/compiler/decorators.ts]: Unrelated to this file, but I'd `console.warn` when an `@component` decorator is encountered in order to make users aware of it being legacy. `@managed` is a MUCH better name.

File: tests/compiler/simple-unit/index.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+interface Interface {
+  test: Test;
+}
+
+export class TestWithDependency implements Interface {
+  constructor(public test: Test) {}
+}
+
+@unit
+export class Unit {
```
Comment (Prior99) [inline tests/compiler/simple-unit/index.ts]: What is the difference between a `@unit` and a `@managed`

**Comment by KnisterPeter** [comment]
> The biggest flaw that I can see is that the `@inject` decorator seems no longer to be necessary which is fine from technical perspective but will make the code less readable (see comment).

Based on feedback of others I decided to simplify the decorators to just `@managed`, so you can use that to declare what is a 'managed' component and as well which properties are 'managed'.
Same goes for `@external` and `@component` are both integrated into `@managed`.

> Will units within a container be able to inject each other?

From a component perspective units are just components with an additional role. So yes you can inject units and let things inject into units.
From an architectural perspective units serve the purpose of structuring your dependencies and host factories via `@provides`.

> Can a container be a unit? Perhaps I have big sub-trees in my project that are managed in a container an are orchestrated withing a bigger container (think: Big sections being active withing pages, sections and pages might load and unload).

A container is a special object (the only one I think). No it cannot be a unit. A container must be abstract versus a unit must be concrete.
A container is the building block of a dependency tree hosting the entry points of the tree. Right now containers are not injectable but yes, that might be extended in the future.

> Lifecycle methods on containers?

Yes, but not in the first iteration. I will definitely add them.

> Support for async lifecycle hooks on units / components?

Same as lifecycles on containers.
I have something in mind where you can be notified when your dependency is ready to be used. It will not be available on the component itself, because constructor injection must be synchronous by javascript design. But register something on container level might be good enought.
This will need some time to find the best api, and therefore I defer it a bit.

> I am not sure if (while definitely being more scalable across bigger projects) this introduces to much of an assumptions about the project using TSDI and am unsure if the barrier of entry isn't raised to high when forcing the user to understand and use the concepts of containers and units.

The main concept change here is to write a container with an entry point. There may be even no entry point at all. So in the most simple case you end up with two lines (e.g. `@container export abstract class Container {}` and `new ContainerImpl()`).
The challenge to communicate the what and why is still to do. But from the users of TSDI rigth now I get the same feedback. There are api quirks which you need to understand and I hope may of them will go away with the compiler.

File: lib/compiler/cli.ts [hunk 4]
```diff
@@ -1,15 +1,57 @@
...
+
+  public async run(): Promise<any> {
+    const { flags } = this.parse(CompilerCommand);
+
+    new Compiler(flags.project).getContainers().forEach(container => {
+      const code = container.generate(flags['out-dir']);
+      if (flags['std-out']) {
+        console.log(code);
+      } else {
+        const fileName = kebabCase(container.implName).replace(/^-/, '');
```
[the remark this answers is not in the document]
Comment (KnisterPeter) [reply lib/compiler/cli.ts]: Good point

File: lib/compiler/component.ts [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+  }[] = [];
+
+  public get name(): string {
+    if (TypeGuards.isInterfaceDeclaration(this.node)) {
+      return this.node.getName();
+    }
+    return this.node.getNameOrThrow();
+  }
+
+  private readonly symbolId = Component.SymbolId++;
```
[the remark this answers is not in the document]
Comment (KnisterPeter) [reply lib/compiler/component.ts]: That should be no problem since its static to this class. A different version (file in the node_modules folder) will be a different class instance and have its own static counter.
In essence, this symbolId is used to create a serializable unique id. Using `uuid` will introduce a new (probably big) dependency and does not make this id more readable.

File: lib/compiler/container.ts [hunk 6]
```diff
@@ -0,0 +1,161 @@
+import { ClassDeclaration, SyntaxKind, TypeGuards } from 'ts-morph';
+import { Compiler } from '.';
+import { container } from '../tsdi';
+import { Component } from './component';
+import { UnitImpl } from './unit';
+import { findDeclarationForIdentifier } from './util';
+
+export class Container<T> {
```
[the remark this answers is not in the document]
Comment (KnisterPeter) [reply lib/compiler/container.ts]: As written in the longer comment below, a `Container` is the host of a dependency tree with entry points to it.

File: lib/compiler/container.ts [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+        return {
+          name: property.getName(),
+          type: new Component(this, node)
+        };
+      }
+      throw new Error('Illegal node type for component: ' + node.print());
+    });
+  }
+
+  private _units?: UnitImpl[];
```
[the remark this answers is not in the document]
Comment (KnisterPeter) [reply lib/compiler/container.ts]: Right now this field suppresses that units are created multiple times. I think it would be better to make units singletons like the components are (components have a singleton check in the constructor and return the singleton instance already existing).

File: lib/compiler/container.ts [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+          throw new Error('Illegal node type for unit: ' + node.print());
+        }
+      });
+    } else {
+      this._units = [];
+    }
+    return this._units;
+  }
+
+  private get internals(): Component[] {
```
[the remark this answers is not in the document]
Comment (KnisterPeter) [reply lib/compiler/container.ts]: Right. `internals` is every managed piece which is not `external` :smile: 
Internals and externals are a fragment of transitioning from version 0.x where externals where components which need to have injections but could not be managed by TSDI.
This concept is now hidden for most of the users of TSDI.

There are cases where this will be relevant.
One of them would be having multiple container as parts of your dependency tree and having `@managed` classes which are not directly integrated in a managed tree of one of the containers. Then users need to flag which container should managed this piece. And to create a meaningful error message this internal-external thing is there.

File: lib/compiler/container.ts [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    const pending = [
+      ...this.entryPoints.map(entryPoint => entryPoint.type),
+      ...this.units.map(unit => unit.unitComponent)
+    ];
+    const list: Component[] = [];
+
+    while (pending.length > 0) {
```
[the remark this answers is not in the document]
Comment (KnisterPeter) [reply lib/compiler/container.ts]: Good catch

File: lib/compiler/container.ts [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+      return filterInternals(
+        all
+          .filter(component => component.by)
+          .filter(component => component.by!.clazz === this.clazz)
+      );
+    }
+    return filterInternals(all);
+  }
+
+  public get components(): Component[] {
```
[the remark this answers is not in the document]
Comment (KnisterPeter) [reply lib/compiler/container.ts]: See my comments below.

I think the concept will get a lot easier now for most cases. This will reduce the entry barrier.
But in complex setups it is way more flexible compared to before.
TSDI still does not assume and structure or architecture besides that you need to use classes. :smile: That may be unpopular in the JS world, but I like it.

File: lib/compiler/container.ts [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    return [...this.internals, ...this.externals];
+  }
+
+  /**
+   * @internal
+   */
+  public get code(): string {
+    return this.generate(__dirname);
```
[the remark this answers is not in the document]
Comment (KnisterPeter) [reply lib/compiler/container.ts]: Could be done later on, but for now I do not care about performance but for feature completeness and simplicity.

File: lib/compiler/container.ts [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+  }
+
+  public importName: string;
+
+  /**
+   * @internal
+   */
+  constructor(public compiler: Compiler, public clazz: ClassDeclaration) {
+    this.importName = `${this.compiler.idGen}_Container`;
```
[the remark this answers is not in the document]
Comment (KnisterPeter) [reply lib/compiler/container.ts]: Do you mean the style of making parameters to properties?

File: lib/compiler/decorators.ts [hunk 13]
```diff
@@ -52,30 +50,46 @@ export function managed(
...
+  if (typeof config !== 'function') {
+    return () => undefined;
+  }
 }
 
-export function initialize(target: object, propertyKey: string): void;
-export function initialize(): MethodDecorator;
-export function initialize(): MethodDecorator | void {
+export function postConstruct(target: object, propertyKey: string): void;
+export function postConstruct(): MethodDecorator;
```
[the remark this answers is not in the document]
Comment (KnisterPeter) [reply lib/compiler/decorators.ts]: I will rename them then. I would as well favor of keeping `initialize` and `destroy` but since these decorators have runtime side-effects from version 0.x I need to have a new pair.
The compiler version is nearly free of a runtime and has very much less overhead during container creation.

File: lib/compiler/runtime.ts [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+      readFileSync(resolvedId).toString(),
+      this.transpileOptions
+    ).outputText;
+    const exports = this.evaluateModule(code, resolvedId);
+
+    this.containerRequireCache[resolvedId] = exports;
+    return this.containerRequireCache[resolvedId];
+  }
+
+  private evaluateModule(
```
[the remark this answers is not in the document]
Comment (KnisterPeter) [reply lib/compiler/runtime.ts]: To create containers directly in the process of the compiler the generated code need to be evaluated. The main purpose for now is to use this in unit tests and its internal api and will not be bundled in a browser version.
Otherwise it would be necessary to either write the code to disk and then start another process to run it or to move the `hacky` evaluation part into the test code. The second way was integrated in the first compiler iteration which will be overwritten with this PR and has had a lot more flaws since it could not access the internals of the container.

So in summary: yes its a hack, but its a controlled hack for special runtime cases like tests.

File: lib/compiler/decorators.ts [hunk 15]
```diff
@@ -1,26 +1,24 @@
+import { addTsdiMarker } from '../marker';
 import { TSDI } from '../tsdi';
 
 export function managed(configuration: { by: any }): ClassDecorator;
 export function managed<T extends { new (...args: any[]): {} }>(
   constructor: T
 ): T;
+export function managed(configuration: {
```
[the remark this answers is not in the document]
Comment (KnisterPeter) [reply lib/compiler/decorators.ts]: The whole set of legacy decorators is supported. There is no need to repeat the decorators here.
The compiler will search your codebase for usages of the legacy decorators.

File: lib/marker.ts [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+    configurable: false,
+    enumerable: false,
+    writable: false
+  });
+}
+
+export function findTsdiMarker(o: any): any {
+  let proto = o;
+  while (proto !== null) {
+    if (Object.getOwnPropertyDescriptor(proto, '__tsdi_marker__')) {
```
[the remark this answers is not in the document]
Comment (KnisterPeter) [reply lib/marker.ts]: Because on the long run I want to get rid of `reflect-metadata`.
The compiler version does not rely on `emitDecoratorMetadata` anymore and therefore you don't need `reflect-metadata`. Which by the way is very heavyweight with 50k.

File: tests/compiler/cli/container.ts [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+export class Dependency {}
+
+@managed
+export class Test {
+  constructor(public dependency: Dependency) {}
+}
+
+@container
+export abstract class Container {
+  public abstract test: Test;
```
[the remark this answers is not in the document]
Comment (KnisterPeter) [reply tests/compiler/cli/container.ts]: The container is a special case. Nothing here is injected but the property `test` here is an entry point into a managed dependecy tree. So its just a generated accessor. And as well abstract properties or methods could have a decorator in typescript.

File: tests/compiler/cli/test.ts (hunk 1 above)
Comment (KnisterPeter) [reply tests/compiler/cli/test.ts]: > I understand what this is needed for, but my guess would be that one will afterwards at project's toplevel layer import the ContainerImpls from the generated code and instanciate them.

Thats intended use. One should instantiate the container by itself since there is no runtime registry of TSDI anymore. You have to do this instead of creating a TSDI instance (this is done inside of the container).
The registering and accessing via factory method is a nice idea though. But then users still need to import the 'container impl file' somewhere in their project.

File: lib/compiler/container.ts [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+          `Unassigned externals [${
+            unassigned[0].name
+          }] is not supported if using multiple containers`
+        );
+      }
+
+      return filterInternals(
+        all
+          .filter(component => component.by)
+          .filter(component => component.by!.clazz === this.clazz)
```
[the remark this answers is not in the document]
Comment (KnisterPeter) [reply lib/compiler/container.ts]: Tried it, but it will not work as expected.
The compiler does not infer that we filtered all components which are a falsy value on the `by` property. That only works if we filter components which are falsy (e.g. undefined).
